File: app-expand/src/main/java/com/zxn/expandapp/widget/ExpandTextView.java
package com.zxn.expandapp.widget;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.StateListDrawable;
import androidx.appcompat.widget.AppCompatTextView;
import android.util.AttributeSet;
import android.view.Gravity;


/**
 * 带有图片的TextView,图片和文字紧挨在一起的效果.
 * Updated by zxn on 2019/6/29.
 */
public class ExpandTextView extends AppCompatTextView {

    private Rect mTextBound = new Rect();
    private int mDrawableGravity;

    public ExpandTextView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
//        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.ExpandTextView);
//        mDrawableGravity = a.getInt(R.styleable.ExpandTextView_drawable_gravity, 0);
//        a.recycle();
    }

    public ExpandTextView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ExpandTextView(Context context) {
        this(context, null);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        Drawable[] drawables = getCompoundDrawables();
        if (drawables != null) {
            Drawable drawableLeft = drawables[0];
            Drawable drawableRight = drawables[2];
            if (drawableLeft != null) {
                // 文本区域
                getPaint().getTextBounds(getText().toString(), 0, getText().length(), mTextBound);
                float textWidth = mTextBound.width();
                int drawablePadding = getCompoundDrawablePadding();
                int drawableWidth = drawableLeft.getIntrinsicWidth();
                float bodyWidth = textWidth + drawableWidth + drawablePadding;

                int color = getTextColors().getColorForState(getDrawableState(), 0);
                getPaint().setColor(color);
                getPaint().setTextSize(getTextSize());

                canvas.save();
                // 平移画布
                canvas.translate((getWidth() - bodyWidth) / 2, (getHeight() - drawableLeft.getIntrinsicHeight()) / 2);
                drawableLeft.draw(canvas);
                canvas.restore();

                // 绘制文本
                float baseline = (getHeight() - getPaint().getFontMetrics().bottom + getPaint().getFontMetrics().top) / 2 - getPaint().getFontMetrics().top;
                canvas.drawText(getText().toString(), (getWidth() - textWidth - (getWidth() - bodyWidth) / 2), baseline, getPaint());
            }else if (drawableRight != null){
                // 文本区域
                getPaint().getTextBounds(getText().toString(), 0, getText().length(), mTextBound);
                float textWidth = mTextBound.width();
                float textHeight = mTextBound.height();
                int drawablePadding = getCompoundDrawablePadding();
                int drawableWidth = drawableRight.getIntrinsicWidth();

                int color = getTextColors().getColorForState(getDrawableState(), 0);
                getPaint().setColor(color);
                getPaint().setTextSize(getTextSize());
                float bodyWidth = textWidth + drawableWidth + drawablePadding;

                switch (mDrawableGravity) {
                    case Gravity.LEFT:
                        // 平移画布
                        canvas.translate(0, getHeight() / 2);
                        // 绘制文本
                        canvas.drawText(getText().toString(), 0, textHeight / 2, getPaint());

                        // 绘制图片
                        Drawable left = drawableRight.mutate();

                        canvas.translate(textWidth + drawablePadding, 0);
                        left.draw(canvas);
                        break;
                    case Gravity.CENTER:
                        // 平移画布
                        canvas.translate((getWidth() - bodyWidth) / 2, getHeight() / 2);
                        // 绘制文本
                        canvas.drawText(getText().toString(), 0, textHeight / 2, getPaint());

                        // 绘制图片
                        Drawable center = ((StateListDrawable) drawableRight).mutate();

                        canvas.translate(textWidth + drawablePadding, 0);
                        center.draw(canvas);
                        break;
                    case Gravity.RIGHT:
                        // 平移画布
                        canvas.translate(getWidth() - bodyWidth, getHeight() / 2);
                        // 绘制文本
                        canvas.drawText(getText().toString(), 0, textHeight / 2, getPaint());

                        // 绘制图片
                        Drawable right = drawableRight.mutate();

                        canvas.translate(textWidth + drawablePadding, 0);
                        right.draw(canvas);
                        break;
                }
            }else {
                super.onDraw(canvas);
            }
        }else {
            super.onDraw(canvas);
        }
    }

//    public void setGravity(int gravity) {
//        this.mDrawableGravity = gravity;
//        invalidate();
//    }
}

File: expand-widget/src/main/java/com/zxn/widget/ExpandTextView.kt
package com.zxn.widget

import android.content.Context
import android.graphics.Canvas
import android.graphics.Rect
import android.graphics.drawable.StateListDrawable
import android.util.AttributeSet
import android.view.Gravity
import androidx.appcompat.widget.AppCompatTextView

/**
 * 带有图片的TextView,图片和文字紧挨在一起的效果.
 *  Updated by zxn on 2020.10.17.
 */
class ExpandTextView @JvmOverloads constructor(context: Context?, attrs: AttributeSet? = null, defStyle: Int = 0) : AppCompatTextView(context!!, attrs, defStyle) {
    private val mTextBound = Rect()
    private val mDrawableGravity = 0
    override fun onDraw(canvas: Canvas) {
        val drawables = compoundDrawables
        if (drawables != null) {
            val drawableLeft = drawables[0]
            val drawableRight = drawables[2]
            if (drawableLeft != null) {
                // 文本区域
                paint.getTextBounds(text.toString(), 0, text.length, mTextBound)
                val textWidth = mTextBound.width().toFloat()
                val drawablePadding = compoundDrawablePadding
                val drawableWidth = drawableLeft.intrinsicWidth
                val bodyWidth = textWidth + drawableWidth + drawablePadding
                val color = textColors.getColorForState(drawableState, 0)
                paint.color = color
                paint.textSize = textSize
                canvas.save()
                // 平移画布
                canvas.translate((width - bodyWidth) / 2, (height - drawableLeft.intrinsicHeight) / 2.toFloat())
                drawableLeft.draw(canvas)
                canvas.restore()

                // 绘制文本
                val baseline = (height - paint.fontMetrics.bottom + paint.fontMetrics.top) / 2 - paint.fontMetrics.top
                canvas.drawText(text.toString(), width - textWidth - (width - bodyWidth) / 2, baseline, paint)
            } else if (drawableRight != null) {
                // 文本区域
                paint.getTextBounds(text.toString(), 0, text.length, mTextBound)
                val textWidth = mTextBound.width().toFloat()
                val textHeight = mTextBound.height().toFloat()
                val drawablePadding = compoundDrawablePadding
                val drawableWidth = drawableRight.intrinsicWidth
                val color = textColors.getColorForState(drawableState, 0)
                paint.color = color
                paint.textSize = textSize
                val bodyWidth = textWidth + drawableWidth + drawablePadding
                when (mDrawableGravity) {
                    Gravity.LEFT -> {
                        // 平移画布
                        canvas.translate(0f, height / 2.toFloat())
                        // 绘制文本
                        canvas.drawText(text.toString(), 0f, textHeight / 2, paint)

                        // 绘制图片
                        val left = drawableRight.mutate()
                        canvas.translate(textWidth + drawablePadding, 0f)
                        left.draw(canvas)
                    }
                    Gravity.CENTER -> {
                        // 平移画布
                        canvas.translate((width - bodyWidth) / 2, height / 2.toFloat())
                        // 绘制文本
                        canvas.drawText(text.toString(), 0f, textHeight / 2, paint)

                        // 绘制图片
                        val center = (drawableRight as StateListDrawable).mutate()
                        canvas.translate(textWidth + drawablePadding, 0f)
                        center.draw(canvas)
                    }
                    Gravity.RIGHT -> {
                        // 平移画布
                        canvas.translate(width - bodyWidth, height / 2.toFloat())
                        // 绘制文本
                        canvas.drawText(text.toString(), 0f, textHeight / 2, paint)

                        // 绘制图片
                        val right = drawableRight.mutate()
                        canvas.translate(textWidth + drawablePadding, 0f)
                        right.draw(canvas)
                    }
                }
            } else {
                super.onDraw(canvas)
            }
        } else {
            super.onDraw(canvas)
        }
    }
}



//    public void setGravity(int gravity) {
//        this.mDrawableGravity = gravity;
//        invalidate();
//    }